### PR TITLE
Disallow some types of sources for parallel ES explicitly

### PR DESCRIPTION
**Pull request by klecki**
## Description
- [x] **Bug fix** (*non-breaking change which fixes an issue*)
- [ ] **New feature** (*non-breaking change which adds functionality*)
- [ ] **Breaking change** (*fix or feature that would cause existing functionality to not work as expected*)
- [ ] **Refactoring** (*Redesign of existing code that doesn't affect functionality*)
- [ ] **Other** (*e.g. Documentation, Tests, Configuration*)

### What happened in this PR
Only one-argument callables in per-sample mode
are supported. Other options do not allow us
to distribute the work between worker processes.

Update docs.

Detect the disallowed cases and provide 
a meaningful error message.
Add test coverage.

Signed-off-by: Krzysztof Lecki <klecki@nvidia.com>
 
#### Additional information
- Affected modules and functionalities:
Parallel External Source

- Key points relevant for the review:
Nah

### Tests
- [ ] Existing tests apply
- [x] New tests added
  - [x] Python tests
  - [ ] GTests
  - [ ] Benchmark
  - [ ] Other
- [ ] N/A

### Documentation
- [ ] Existing documentation applies
- [x] Documentation updated
  - [x] Docstring
  - [ ] Doxygen
  - [ ] RST
  - [ ] Jupyter
  - [ ] Other
- [ ] N/A

#### Requirements
- [ ] Implements new requirements
- [ ] Affects existing requirements
- [x] N/A

**REQ IDs**: N/A

**JIRA TASK**: DALI-2213

### CODE REVIEW

**Comment by klecki** [comment]
!build

File: dali/python/nvidia/dali/external_source.py [hunk 1]
```diff
@@ -341,6 +342,15 @@ class ExternalSource():
 
     When ``parallel`` is set to True, ``source`` must return NumPy/MXNet/PyTorch CPU array,
     TensorCPU, or tuple/list of these types with length matching num_outputs.
+
+    Only callables that accept one argument (:meth:`~nvidia.dali.types.SampleInfo` objects that
+    represent the index of the requested sample) can be used as ``source`` when ``parallel`` is
+    set to True. It can be a function or an object implementing ``__call__`` operator, which
+    allows to add some initial state to the object instance.
```
Comment (JanuszL) [inline dali/python/nvidia/dali/external_source.py:349]:
```suggestion
    allows to add an initial state to the object instance.
```

File: dali/python/nvidia/dali/external_source.py [hunk 2]
```diff
@@ -464,6 +475,20 @@ def __call__(
                 raise ValueError(
                     "``prefetch_queue_depth`` must be a positive integer, got {}.".format(
                         prefetch_queue_depth))
+            if source_desc.kind == _SourceKind.CALLABLE:
+                if not source_desc.has_inputs:
```
Comment (JanuszL) [inline dali/python/nvidia/dali/external_source.py:479]: What if it accepts more than one argument?

Comment (klecki) [reply dali/python/nvidia/dali/external_source.py:479]: Than it's an error either way, but via the External Source - it only accepts 0 and 1-argument callables.

Comment (JanuszL) [reply dali/python/nvidia/dali/external_source.py:479]: Understood.

**Review by JanuszL — APPROVED** [review]

**Review by awolant — APPROVED** [review]

**Comment by klecki** [comment]
!build

**Pull request merged**